### ISSUE TITLE
Tweaks to phet-io validation

### ISSUE THREAD

**Issue by chrisklus** [phetsims/tandem#191]
In discussion with @samreid and @zepumph, we think that phet-io API validation and tandem validation would be clearer and easier to use if they were more related to each other. We think the best starting approach to this cleanup is having validation be either all on or all off, as the multiple levels of validation feels unnecessary.

Some changes:

- [x] Convert `Tandem.errorOnFailedValidation` to a constant like `Tandem.VALIDATION`. This will also replace `VALIDATE_TANDEMS`. `Tandem.VALIDATION` should be false if `printMissingTandems` is true.
- [x] Look through usages of `Tandem.PHET_IO_ENABLED` and `PhetioObject.isPhetioInstrumented` to see which ones can be replace with `Tandem.VALIDATION` instead.
- [x] Removed the QP `phetioValidateTandems` and replace it with `phetioValidation` as a boolean that defaults to true. If not provided, the package.json flag mentioned below should be the next place to determine the value of `Tandem.VALIDATION`.
- [x] Update phetmarks to reflect these query parameter changes. We think `phetioValidation` should not be added to the URL by default like `phetioValidateTandems` currently is. A three-state radio button may be a nice solution, where `simulation default`, `true`, or `false` are the options. More on this in https://github.com/phetsims/phetmarks/issues/37.
- [x] Add `ea` to URL in generatePhetioAPI.js (will be in https://github.com/phetsims/chipper/commit/beaa4fd18398d2842d46edf320089e194720f32c )
- [x] PhetioAPIValidation.enabled should rely on Tandem.VALIDATING.
- [x] Create a package.json flag called `validation`. This determines whether the sim will run with validation when it starts up. Query parameters can be added when running locally to temporarily flip the state of validation in the sim, but the default is determined by this package.json flag. It has no default and is required when the sim is run (checked by an assertion in a phet-io related place, like Tandem.js)
- [x] Add validation flag to every phet-io sim's package at phet-io.validation.
- [x] Omit testable-phet-io-validated and remove its usage from listContinuousTests.js.
- [x] Update rc.js so that `"validation": false` is not allowed in a publication.
- [x] Update docs to clearly identify how to turn off validation in a sims package.json so that newly instrumented sims don't fail on CT.

**Comment by samreid** [comment]
I'm planning to start on this today, not sure how far I will get.

**Comment by samreid** [comment]
I tested on a clean working copy by adding an uninstumented `Checkbox` and running with `?phetioValidateTandems=true` and was surprised to see that this was not caught.  I wonder if I'm testing wrong or if this will need to be corrected.

**Comment by samreid** [comment]
I split #192 into a separate issue and will continue here.

**Comment by samreid** [comment]
To test my working copy changes, I added this code in GravityAndOrbitsModel.js constructor:

```js
    const p = new Property( true, {
      tandem: tandem.createTandem( 'hello' )
    } );
```

Note the missing `phetioType` which is required for VALIDATION.

I tested the following cases:

package.json validation | query parameter
true | omitted
true | true
true | false
false | omitted
false | true
false | false

Each case worked as expected.  I think I'll commit and reach out to @zepumph for a review.

**Comment by samreid** [comment]
Repeating a comment from earlier since all the commits pushed it up too far:

To test my working copy changes, I added this code in GravityAndOrbitsModel.js constructor:

```js
    const p = new Property( true, {
      tandem: tandem.createTandem( 'hello' )
    } );
```

Note the missing `phetioType` which is required for VALIDATION.

I tested the following cases:

package.json validation | query parameter
true | omitted
true | true
true | false
false | omitted
false | true
false | false

Each case worked as expected.  I think I'll commit and reach out to @zepumph for a review.

This is ready for review, in particular, be aware of:
* https://github.com/phetsims/tandem/issues/192
- [ ] Please skim through usages of `Tandem.VALIDATION` and make sure they are used appropriately.
- [ ] Please skim through usages of `Tandem.PHET_IO_ENABLED` and make sure they are used appropriately
- [ ] Test in a few conditions, keep in mind you have to `grunt update` after changing package.json
- [ ]  Review the changes in the tandem repo
- [ ] Review the changes in the phet-io repo
- [ ] Test the new radio buttons in phetmarks
- [ ] Note that the tandem repo has validation: false in package.json.  There was some interference with unit tests.  How do you recommend to proceed on that?

**Comment by samreid** [comment]
I'm seeing some CT issues that may be related.  Self-assigning.

**Comment by samreid** [comment]
The problem is that listContinuousTests.js says:

```js
  // All tests should work with no query parameters, with assertions enables and also in phet-io brand
  [ '', '?ea', '?brand=phet-io', '?ea&brand=phet-io' ].forEach( queryString => {
    if ( repo === 'phet-io' && !queryString.includes( 'phet-io' ) ) {
```

But many repos, such as "phetcommon" don't list "phet-io" as a supported brand and hence fail when they are missing the package.json "phet"."phet-io"."validation" attribute.

I'll commit a change so that phet-io brand is only specified for phet-io repos.

**Comment by samreid** [comment]
I asked slack dev-public:

> Do I need to restart CT after changes to listContinuousTests.js? 

I'll check CT in an hour or so to see what happens.

**Comment by samreid** [comment]
I pushed fixes for some straggling issues on CT, these should be reviewed as well.  I wasn't sure that dot/kite/scenery should pick up a dependency on Query String Machine, but this seemed straightforward and will be superseded when QueryStringMachine becomes a module.

**Comment by zepumph** [comment]
I have not yet conducted any review of this issue, but I do want to note that I have hit to following error three times in the last 3 work days:

`Uncaught Error: Assertion failed: phet-io is a required attribute in package.json for brand=phet-io`

All of which are trying to run phet-io brand in a sim/demo/unit test that isn't in `perennial/data/phet-io`. I have two thoughts on this:

1. Either add this flag to every single package.json flag in the project, and set `validation` to `false` for those not in `phet-io` and likely for some inside that list.
2. Make this flag optional, with a default. I know we talked about doing this during our discussion, but I don't remember the negatives of it fully. Whatever they are, currently I feel like the current workflow is no scalable: needing to temporarily add a flag to package.json, `grunt update`, and then re-run the failing page in phet-io brand.

**Comment by samreid** [comment]
Repos that we run with `?brand=phet-io` should have `supportedBrands:['phet-io']` and a `phet-io` sections in package.json.  I recommend that we add these on-demand as needed instead of adding this to every repo eagerly.

**Comment by zepumph** [comment]
That workflow is annoying for me for the kind of work I'm doing often, but likely not for many others. I'm happy to adhere if that is the consensus.

**Comment by samreid** [comment]
Let's ask for feedback at a developer meeting, the proposed strategies are:

Strategy A: In order to run in phet-io brand, a repo's package.json must have a "phet-io" section that indicates "validation": true or false
Strategy B: Any repo can be run in phet-io brand, and if there is no "phet-io"."validation", it would default to "validation":true.  In this strategy, "validation" would only be specified in package.json as false as an opt-out.

**Comment by jonathanolson** [comment]
Woah, looks like QueryStringMachine was added as an external dependency for running Scenery? If it's a requirement, we should be bundling that type of thing into all of our builds. (It went from "have lodash/jquery" to "have lodash/jquery + QueryStringMachine" for anyone using our libraries.

**Comment by samreid** [comment]
I did add that as a dependency, the usage is in `Tandem.js`.  I wasn't sure whether that was best, or if I should find a workaround for this.  What if in the future QueryStringMachine is an es6 module?  Would that be OK or still problematic?

**Comment by jonathanolson** [comment]
> What if in the future QueryStringMachine is an es6 module?

Preferred, if we don't need it in non-module contexts. Or included in the standalone builds.

**Comment by zepumph** [comment]
For my review (developer meeting ignore this):
* It is possible that some test html files were updated, for example I hit a validation flag change while calling grunt update for studio-tests.html over in https://github.com/phetsims/studio/issues/171

**Comment by zepumph** [comment]
@jonathanolson, would the dependency be considered squashed with the following patch? This is potentially as a temporary work-around until we can support QSM as a module (many non-sim usages need QSM at the moment).

```diff
Index: js/Tandem.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/Tandem.js	(revision 9810b19e139279066337077d4d23f23c61144adb)
+++ js/Tandem.js	(date 1595552033259)
@@ -28,7 +28,7 @@
 
 // By default, package.json's value for "phet"."phet-io"."validation" determines whether validation is enabled,
 // but this can be overridden with a query parameter ?phetioValidation={true|false}.
-const IS_VALIDATION_QUERY_PARAMETER_SPECIFIED = QueryStringMachine.containsKey( 'phetioValidation' );
+const IS_VALIDATION_QUERY_PARAMETER_SPECIFIED = QueryStringMachine ? QueryStringMachine.containsKey( 'phetioValidation' ) : false;
 const IS_VALIDATION_SPECIFIED = ( PHET_IO_ENABLED && IS_VALIDATION_QUERY_PARAMETER_SPECIFIED ) ? phet.preloads.phetio.queryParameters.phetioValidation :
                                 ( PHET_IO_ENABLED && packageJSON.phet[ 'phet-io' ].validation );

**Comment by jonathanolson** [comment]
Would need to be `window.QueryStringMachine` I believe, but that would be fine with me (then QSM could be removed from most of those files that it was added to that would never use phet-io?)

**Comment by zepumph** [comment]
Agreed! That sounds really nice. I'll note that as part of my review for the issue.

**Comment by zepumph** [comment]
Review is complete @samreid, please see https://github.com/phetsims/tandem/issues/191#issuecomment-663259433. We will still discuss https://github.com/phetsims/tandem/issues/191#issuecomment-662859961 in dev meeting.

**Comment by zepumph** [comment]
During dev meeting discussion today, we decided that validation should be 

CM said he slightly preferred strategy B to A, but didn't have a strong preference, and that it would be mostly based on what was easiest in the code.

JB: can we make it consistent with out package.json flags, where `validation` defaults to false, and you opt in just like all other parameters.

CK: brought up that to achieve the opt-in behavior, we could name it like `validationOptOut`: true

Let's bring this up back in phet-io dev meeting.

**Comment by samreid** [comment]
> It is possible that some test html files were updated, for example I hit a validation flag change while calling grunt update for studio-tests.html over in phetsims/studio#171

Thanks, I have subsequently run `grunt update` on everything for another issue.  Is there more to do here?

> ?phetioValidation has a defaultValue that we never intend to use. I think it would be prefereable to make the default value null, then the line in Tandem can be a be consolidated:

Can a `type: boolean` query parameter have a default value of `null`?

> then the line in Tandem can be a be consolidated:
> Or maybe some other way that provides proper precedent to the query parameter.

Should it use `containsKey` instead?

> If the QSM usage sticks around, please remove all usages of adding it as a dependency, since we made it graceful in 1d79940. I mean reverting commits like phetsims/dot@0b74e22

Yes, I should do that, thanks!

> In Tandem.js, there seems to be no reason to stringify packageString just to parse it, can't you just return either object as the packageJSON?

Great idea, I'll investigate.

**Comment by samreid** [comment]
I made a few commits and added questions and comments above, back to @zepumph for re-review.

**Comment by zepumph** [comment]
>Can a type: boolean query parameter have a default value of null?


Looks like not. I played around with a custom boolean/null type like so:
```js
    phetioValidation: {
      type: 'custom', defaultValue: null,
      parse: x => x === 'true' ? true : x === 'false' ? false : x,
      isValidValue: value => typeof value === 'boolean' || value === null
    },
```

It worked, but did not actually make the Tandem.js usage any better. My recommendation is to keep it. The only issue I can think of running into down the road is if we eventually set up wrapper side code to always provide the query parameter, like our pattern for `phetioEmitHighFrequencyEvents`. Would there ever be a case where we provided it via the query parameter like that but we wanted to read from packageJSON? I don't think we need to worry about this right now; just brainstorming.


>JB: can we make it consistent with out package.json flags, where validation defaults to false, and you opt in just like all other parameters.

I feel like this was a direction I wanted to discuss with the phet-io team sometime soon. It feels like the best way to add this feature. That said, we want validation on by default, so we may need to work it out differently. 



>Strategy C? Any repo can be run in phet-io brand, but validation is still a required parameter to the sim. If it is not provided by a package.json flag, then a query parameter ?phetioValidation={{true}} must be provided. This is similar to strategy A in which we don't have the confusion of a default value

I think that we should only assert that the package.json flag is present when the query parameter is not supplied. This will help interim development in sims that may not have phet-io as a supported brand just yet.

**Comment by samreid** [comment]
How about if we default validation to true & it doesn't need to be specified in package.json.  The only allowed value in package.json is "validation: false"?  We would never assert that the package.json flag is provided.  I don't want the validation to default to false, or to have to visit every package.json and add "validation:true" to it once it is ready.

If that sounds agreeable, I can proceed on that.  Otherwise we should probably schedule for Wednesday discussion before I continue.

**Comment by zepumph** [comment]
I think that is the appropriate solution. It is opposite from all the other "opt in" flags in package.json, but that is the only con I can think of. I think in practice, only the few sims currently being worked on in initial development (less than 5?) will opt out of validation. So it makes the most sense to set things up sparsely like that. 

- [ ] Please also make sure to update the instrumentation guide appropriately to communicate this opt-out flag to be used in initial development, as well as a place to remind devs to turn validation back on.

**Comment by samreid** [comment]
I recommend dividing the remaining work into side issues, I'll start at the review comment https://github.com/phetsims/tandem/issues/191#issuecomment-663259433

**Comment by samreid** [comment]
OK I think I moved all remaining work to side issues.  @zepumph can you please review this original issue to see if it can be closed?
